### DISCUSSION TITLE
Imagemagick optimisation 1 command

### DISCUSSION THREAD

**Discussion by fabienklesse** [ImageMagick/ImageMagick#3578] [Development]
Hi everyone,
I have a problem, I would like to optimize the time for creating an image. Currently I am opening a convert for each step but I am sure there is better to do.

exec("/usr/bin/convert -quality 100 ".$dessus." -background transparent -distort SRT ".($aiguilleheure-45)." -matte -virtual-pixel transparent -distort Perspective '0,0 579,6 ".imagesx($dessuscoordonnees).",0 708,140 ".imagesx($dessuscoordonnees).",".imagesy($dessuscoordonnees)." 585,265 0,".imagesy($dessuscoordonnees)." 452,126' +level 2%,100% ".$dessus_distordu); 

exec("/usr/bin/convert -quality 100 ".$final." -page +0+0 ".$dessus_distordu." -flatten ".$final);

Is there a way to put everything in the same convert command, without writing the intermediate files to disk?
I've try that but it doesn't work:

exec("/usr/bin/convert -quality 100 ".$final." -page +0+0 (".$dessus." -background transparent -distort SRT ".($aiguilleheure-45)." -matte -virtual-pixel transparent -distort Perspective '0,0 579,6 ".imagesx($dessuscoordonnees).",0 708,140 ".imagesx($dessuscoordonnees).",".imagesy($dessuscoordonnees)." 585,265 0,".imagesy($dessuscoordonnees)." 452,126' +level 2%,100%) -flatten ".$final);

Thanks a lot 
Best
Fabien

**Comment by snibgo** [comment]
Your try at a combined command looks reasonable, but I can see two problems:

1. The parentheses both need a space before and a space after.

2. "-flatten" uses the current background. This defaults to "white". In your combined command, you set this to "transparent". If you need "white", then set "-background white" before "-flatten".

**Reply by fabienklesse** [reply]
Thanks snibgo but doesn't work like that with spaces before and after parentheses.
All what is between parentheses is not created, i see that because the page loads in 7.5 seconds and before it was 9 secondes with 2 converts

exec("/usr/bin/convert -quality 100 ".$final." -page +0+0 ( ".$dessus." -background transparent -distort SRT ".($aiguilleheure-45)." -matte -virtual-pixel transparent -distort Perspective '0,0 579,6 ".imagesx($dessuscoordonnees).",0 708,140 ".imagesx($dessuscoordonnees).",".imagesy($dessuscoordonnees)." 585,265 0,".imagesy($dessuscoordonnees)." 452,126' +level 2%,100% ) -flatten ".$final);

**Comment by snibgo** [comment]
What version of IM?

Please show the actual command used, without variables. You could build the command into a string variable, then echo the value of the variable, and "exec" it.

Please also show the input images.

**Reply by fabienklesse** [reply]
Here is my code, I don't know the version, it's on a ionos server:

the generated picture is here, if i send 10800 , the number is 1 and the clock shows 8 o'clock
https://www.flitch.me/predictionarchives/1396010813_1619183300.jpg

```
// STATION DEBUT

if ($tapis=="station"){

$dessous='lestapispersos/'.$tapis.'.jpg';
$dessus='lestapispersos/aiguilles/aiguilleheuregare.png';
$dessus2='lestapispersos/aiguilles/aiguilleminutegare.png';

$dessus_distordu='distort/'.$tapis.$flitchid.'.png';
$resize="distort/resize".$flitchid.".png";
$final='distort/distortfinal_'.$flitchid.'.jpg';
$dessuscoordonnees = imagecreatefrompng($dessus);
$dessouscoordonnees = imagecreatefromjpeg($dessous);
$largeurbackground=imagesx($dessouscoordonnees);
$hauteurbackground=imagesy($dessouscoordonnees);

$dessuscarte=substr($texte,0,1);

//---------------------------------------
$destinationpng = imagecreatefrompng("lestapispersos/trans50portrait.png");
imagealphablending($destinationpng, true);
imagesavealpha($destinationpng, true);

$color = imagecolorallocatealpha($destinationpng, 0, 0, 10, 20);

if ($dessuscarte==1){$dechorloge=-180;}else{$dechorloge=-140;}
imagettftext($destinationpng, $tailletexte+1400, 0+$inclinaison, $dechorloge+$tx+$decgd+($aligne*1), 1350+$ty+$dechb+($interligne*1), $color, $font, stripslashes($dessuscarte));


imagepng($destinationpng, $resize);
exec("/usr/bin/convert -quality 100 ".$resize." -resize ".$largeurbackground."x".$hauteurbackground."\! -matte -virtual-pixel transparent -distort Perspective '0,0 293,75 ".imagesx($dessouscoordonnees).",0 440,83 ".imagesx($dessouscoordonnees).",".imagesy($dessouscoordonnees)." 441,272 0,".imagesy($dessouscoordonnees)." 291,264' ".$dessus_distordu);
exec("/usr/bin/convert -quality 100 ".$dessous." -page +0+0 ".$dessus_distordu." -flatten ".$final);

//-------------

//exec("/usr/bin/convert -font Bebas.ttf -background transparent -size ".$largeurbackground."x".$hauteurbackground."  -gravity west -pointsize 220 -fill black caption:'salut vous êtes sur flitch' +level 5%,100% -matte -virtual-pixel transparent -distort Perspective '0,0 293,75 ".imagesx($dessouscoordonnees).",0 440,83 ".imagesx($dessouscoordonnees).",".imagesy($dessouscoordonnees)." 441,272 0,".imagesy($dessouscoordonnees)." 291,264' +level 10%,100% ".$resize);
//exec("/usr/bin/convert -quality 100 ".$dessous." -page +0+0 ".$resize." -flatten ".$final);


$aiguilleminute=(substr($texte,3,2)*6);
$aiguilleheure=(substr($texte,1,2)*30)+(substr($texte,3,2)/2);

// Ombres
exec("/usr/bin/convert -quality 100 ".$dessus2." -background transparent -distort SRT ".($aiguilleminute-45)." -blur 0x20 -matte -virtual-pixel transparent -distort Perspective '0,0 579,6 ".imagesx($dessuscoordonnees).",0 708,140 ".imagesx($dessuscoordonnees).",".imagesy($dessuscoordonnees)." 585,265 0,".imagesy($dessuscoordonnees)." 452,126' +level 15%,100% ".$dessus_distordu); 

// METTRE $dessous UNIQUEMENT SUR LE PREMIER... SINON $final
exec("/usr/bin/convert -quality 100 ".$final." -page +1+3 ".$dessus_distordu." -flatten ".$final);


exec("/usr/bin/convert -quality 100 ".$dessus." -background transparent -distort SRT ".($aiguilleheure-44)." -blur 0x20 -matte -virtual-pixel transparent -distort Perspective '0,0 579,6 ".imagesx($dessuscoordonnees).",0 708,140 ".imagesx($dessuscoordonnees).",".imagesy($dessuscoordonnees)." 585,265 0,".imagesy($dessuscoordonnees)." 452,126' +level 15%,100% ".$dessus_distordu); 

// METTRE $dessous UNIQUEMENT SUR LE PREMIER... SINON $final
exec("/usr/bin/convert -quality 100 ".$final." -page +3+3 ".$dessus_distordu." -flatten ".$final);
// Fin Ombres



exec("/usr/bin/convert -quality 100 ".$dessus." -background transparent -distort SRT ".($aiguilleheure-45)." -matte -virtual-pixel transparent -distort Perspective '0,0 579,6 ".imagesx($dessuscoordonnees).",0 708,140 ".imagesx($dessuscoordonnees).",".imagesy($dessuscoordonnees)." 585,265 0,".imagesy($dessuscoordonnees)." 452,126' +level 2%,100% ".$dessus_distordu); 

// METTRE $dessous UNIQUEMENT SUR LE PREMIER... SINON $final
exec("/usr/bin/convert -quality 100 ".$final." -page +0+0 ".$dessus_distordu." -flatten ".$final);

exec("/usr/bin/convert -quality 100 ".$dessus2." -background transparent -distort SRT ".($aiguilleminute-45)." -matte -virtual-pixel transparent -distort Perspective '0,0 579,6 ".imagesx($dessuscoordonnees).",0 708,140 ".imagesx($dessuscoordonnees).",".imagesy($dessuscoordonnees)." 585,265 0,".imagesy($dessuscoordonnees)." 452,126' +level 2%,100% ".$dessus_distordu); 

// METTRE $dessous UNIQUEMENT SUR LE PREMIER... SINON $final
exec("/usr/bin/convert -quality 100 ".$final." -page +0+0 ".$dessus_distordu." -flatten ".$final);


//exec("/usr/bin/convert -font vavont.ttf -background transparent -size ".$largeurbackground."x".$hauteurbackground." -gravity Center -pointsize 120 -fill blue caption:'".$texte."' +level 60%,100% ".$resize);
//exec("/usr/bin/convert -quality 100 ".$final." -page +0+0 ".$resize." -flatten ".$final);

//------------------------



$a++;$b++;
$source = imagecreatefromjpeg($final);
$largeur_source = imagesx($source);
$hauteur_source = imagesy($source);
imagealphablending($source, true);
imagesavealpha($source, true);
$destination = imagecreatefromjpeg($final);
$largeur_destination = imagesx($destination);
$hauteur_destination = imagesy($destination);
$destination_x = 0;
$destination_y =  0;
imagecopy($destination, $source, $destination_x, $destination_y, 0, 0, $largeur_source, $hauteur_source);
imagejpeg($destination, "predictiontemp/".$ithumpid.".jpg");
imagejpeg($destination, "predictionarchives/".$ithumpid."_".$heure.".jpg");


$imgpourbase="predictionarchives/".$ithumpid."_".$heure.".jpg";

if (strlen($textekeepetnouveau)>=5){
$req="DELETE FROM vb_keep WHERE ithumpid='$ithumpid'";
mysql_query($req);$videbase=1;}
}
// STATION FIN
```

**Reply by fabienklesse** [reply]
here is the concat line that doesn't work:
/usr/bin/convert -quality 100 distort/distortfinal_466803068d30f66b84fc2bac1a07e8e2.jpg -page +0+0 (lestapispersos/aiguilles/aiguilleheuregare.png -background transparent -distort SRT 255 -matte -virtual-pixel transparent -distort Perspective '0,0 579,6 2000,0 708,140 2000,2000 585,265 0,2000 452,126' +level 2%,100%) -flatten distort/distortfinal_466803068d30f66b84fc2bac1a07e8e2.jpg

**Comment by snibgo** [comment]
You need a space after open parenthesis in "( lestapispersos/...", and before close parenthesis in "+level 2%,100% )".

Please show your input images distortfinal_466803068d30f66b84fc2bac1a07e8e2.jpg and aiguilleheuregare.png.

**Reply by fabienklesse** [reply]
Here are the 2 pics 👍 
https://flitch.me/distort/distortfinal_466803068d30f66b84fc2bac1a07e8e2.jpg

https://flitch.me/lestapispersos/aiguilles/aiguilleheuregare.png

doesn't work with the spaces...
thank you for the time you spend with me
